diff --git a/modules/filefield_paths.inc b/modules/filefield_paths.inc
index f211b8d..f07d731 100644
--- a/modules/filefield_paths.inc
+++ b/modules/filefield_paths.inc
@@ -48,7 +48,7 @@ function filefield_paths_filefield_paths_process_file($type, $entity, $field, $i
   if (isset($instance['settings']['filefield_paths'])) {
     $settings = $instance['settings']['filefield_paths'];
     foreach ($items as &$file) {
-      if (filefield_paths_uri_is_local($file['uri']) && ($file['timestamp'] == REQUEST_TIME || $settings['active_updating'])) {
+      if (filefield_paths_uri_is_local($file['uri'])) {
         $token_data = array(
           'file' => (object) $file,
           $type => $entity
